### PR TITLE
Queues - Andrea Valliere - Ride-Share-Two

### PR DESCRIPTION
**Pull request by avalliere**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    | Instructor Feedback
| :------------- | :------------- |:---|
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |    For the Trip class, I initially wanted to have one find_many(id) class method that would be simple enough it could work for either  finding many instances of Trip by driver_id or by rider_id. With the way I started out writing the code, it ended up being two separate methods for find_many_drivers and find_many_riders. I still think it may have been better to create a more versatile find_many method, but I wasn't able to figure out how to implement it.  |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  Class methods are making more sense to me now, after this project and Scrabble. I'd like to add that I feel less clear about test coverage and hope I could get some specific feedback about increasing coverage percentage. I have a low test coverage and as I added more detailed tests, the coverage percentage went down even lower.  | That is odd - I wouldn't expect that adding tests would ever decrease coverage. Make sure you're starting simplecov at the very top of `spec_helper.rb`, before even `require`-ing any of the minitest files. |
| Describe a _nominal test_ that you wrote for this assignment. |  Most of my class methods begin with very general nominal tests to check that I am getting the data type that I expect. For instance, with the trips method in my Driver class, I expect to have an array returned of all trips for one specific Driver. My first test is just to see that I am getting an array.  |
| Describe an _edge case test_ that you wrote for this assignment. |  For my find class method in Driver class, I have tested to see that the name and driver ID match what I expect for the first and last drivers in the database.  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I began with a chart of interactivity between my classes, and this actually threw me off and made me feel overwhelmed. Once I started writing code (and first writing pseudocode on a white board) I felt a lot less confused. This has made me realize that I have a greater level of comfort with the details of a program and less with the broader concepts of interactivity throughout my programs - something to work on.  |

### CODE REVIEW

File: specs/driver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
+require_relative 'spec_helper'
+
+describe "Driver class" do
+  describe "Initialize method" do
```
Comment (droberts-sea) [inline specs/driver_spec.rb:4]: Good organization - the nested `describes` make this code much easier to read.

File: specs/driver_spec.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    end
+  end
+
+  describe "trips method" do
+    it "Retrieves the list of trip instances unique to one instance of driver" do
+      # skip
+      bernardo_trips = RideShare::Driver.find("1").trips
+      bernardo_trips.must_be_instance_of Array
+      bernardo_trips.length.must_equal 9
+    end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:28]: I would like to see some more coverage here around the contents of the array you get from `#trips`. For example:
- Are all the elements you got back instances of `Trip`?
- Do all the elements you got back have this driver as their `driver_id`?

File: lib/rider.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      @name = name
+      @phone_num = phone_num
+    end
+
+    def trips
+      Trip.find_many_riders(@rider_id)
+    end
+
+    def drivers
+      # Currently just returning IDs for drivers. May want to refactor to return names of drivers?
```
Comment (droberts-sea) [inline lib/rider.rb:17]: You will probably want to return the driver objects themselves - much more useful than just an ID.

File: lib/rider.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      drivers = []
+      rider_trips = Rider.find(@rider_id).trips
```
Comment (droberts-sea) [inline lib/rider.rb:19]: Since this method is already inside the `Rider`, why not just call `trips` directly here? As in,

```ruby
rider_trips = trips
```

File: lib/trip.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        trips << self.new(line[0], line[1], line[2], line[3], line[4].to_f)
+      end
+      return trips
+    end
+
+    def self.find_many_drivers(id)
+      driver_trips = []
+      trips = self.all
+
+      trips.each do |trip|
```
Comment (droberts-sea) [inline lib/trip.rb:26]: Could you simplify this logic using an enumerable? Perhaps `select`?

File: lib/trip.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  class Trip
+    attr_accessor :trip_id, :driver_id, :rider_id, :date, :rating
+
+    def initialize(trip_id, driver_id, rider_id, date, rating)
+      @trip_id = trip_id
+      @driver_id = driver_id
+      @rider_id = rider_id
+      @date = date
+      @rating = rating
+    end
```
Comment (droberts-sea) [inline lib/trip.rb:12]: You're missing a couple of methods here! `Trip#rider` and `Trip#driver` should return the `Rider` and `Driver` instances associated with this trip.

File: specs/trip_spec.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      trip.rating.must_equal rating
+      # Needs test to see that rating is between 1 and 5
+    end
+  end
+
+  describe "self.find_many_drivers class method" do
+    it "Returns array of many Trip instances for given driver" do
+      driver_trips = RideShare::Trip.find_many_drivers("1")
+      driver_trips.must_be_instance_of Array
+      driver_trips.length.must_equal 9
```
Comment (droberts-sea) [inline specs/trip_spec.rb:36]: As in driver_spec.rb, I would like to see more testing around the elements in the returned array. What type of thing did you get back? Is the data what you expected to see?

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
|   Used Git Regularly	|   yes - good commit messages	|
|   Answer comprehension questions	|   yes - see responses inline	|
|  **Driver** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   mostly - see inline comments |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips list to calculate the average rating | yes |
|  **Rider** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |  yes  |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips method to retrieve the associated drivers | mostly good - see inline comments |
|  **Trip** |   |
|   Reads the CSV file in the `all` method | yes |
|   Has appropriate edge-case tests for each method in the class |   yes |
|   Created a method that uses a method from the `Driver` to retrieve the associated driver instance  |   missing |
|   Created a method that uses a method from the `Rider` to retrieve the associated rider instance |   missing |
|   Created a method to retrieve all trips by driver id |    yes |
|   Created a method to retrieve all trips by rider id |     yes |
|   **Overall**	|   	|

Great work overall! You're missing a couple of methods in `Trip` and your test coverage could be a little better in places, but on the whole I'm quite happy with this code. Keep up the good work!
